snakemake/persistence.py

```diff
@@ -231,6 +231,8 @@ def finished(self, job, keep_metadata=True):
         fallback_time = time.time()
         for f in job.expanded_output:
             rec_path = self._record_path(self._incomplete_path, f)
+            if not os.path.exists(rec_path):
+                rec_path = self._record_path(self._metadata_path, f)
```

Comment:
I think this addresses the problem of calling finished twice. 
However, small remaining issue is that after the second 'finished', I believe this will give an incorrect starttime equal to the endtime of the job, as the flow becomes:

finished 1:
write metadata-tracker file (with correct starttime from creation time of incomplete-tracker file, etc).
remove incomplete-tracker file

finished 2:
do not find incomplete-tracker file, use metadata-tracker file as source.
write new metadata-trakcer file, but now with incorrect starttime from creation time of metadata-tracker file. 

To solve this, in finished 2 you could parse (_read_record) the metadata-tracker file, and obtain the starttime from it.

Reply:
> To solve this, in finished 2 you could parse (_read_record) the metadata-tracker file, and obtain the starttime from it.

I'm happy to make that change, but I'm a bit unclear at the moment as to how to do it. Could you post some suggested code? I'll also start poking around, but I'm not familiar with the internals here, so I'm a bit lost.

Thanks for talking a look at this. I would agree that if `finished` is being called twice, that is likely an underlying problem (and should probably be solved). However, I'm not sure how that is happening, and thus I'm not able to debug at the moment.